decompiler/structures/visitors/substitute_visitor.py

```diff
@@ -0,0 +1,152 @@
+
+        return self._visit_operation(op)
+
+    def visit_binary_operation(self, op: BinaryOperation) -> Optional[DataflowObject]:
+        return self._visit_operation(op)
+
+    def visit_call(self, op: Call) -> Optional[DataflowObject]:
+        if (function_replacement := op.function.accept(self)) is not None:
+            op._function = _assert_type(
+                function_replacement,
+                Union[FunctionSymbol, ImportedFunctionSymbol, IntrinsicSymbol, Variable]
+            )
+
+        return self._visit_operation(op)
+
+    def visit_condition(self, op: Condition) -> Optional[DataflowObject]:
+        return self._visit_operation(op)
+
+    def visit_ternary_expression(self, op: TernaryExpression) -> Optional[DataflowObject]:
+        return self._visit_operation(op)
+
+    def visit_comment(self, instr: Comment) -> Optional[DataflowObject]:
+        return self._mapper(instr)
+
+    def visit_assignment(self, instr: Assignment) -> Optional[DataflowObject]:
+        if (value_replacement := instr.value.accept(self)) is not None:
+            instr._value = _assert_type(value_replacement, Expression)
+        if (destination_replacement := instr.destination.accept(self)) is not None:
+            instr._destination = _assert_type(destination_replacement, Expression)
+
+        return self._mapper(instr)
+
+    def visit_generic_branch(self, instr: GenericBranch) -> Optional[DataflowObject]:
+        if (condition_replacement := instr.condition.accept(self)) is not None:
+            instr._condition = _assert_type(condition_replacement, Expression)
+
+        return self._mapper(instr)
+
+    def visit_return(self, instr: Return) -> Optional[DataflowObject]:
+        if (values_replacement := instr.values.accept(self)) is not None:
+            instr._values = _assert_type(values_replacement, ListOperation)
+
+        return self._mapper(instr)
+
+    def visit_break(self, instr: Break) -> Optional[DataflowObject]:
+        return self._mapper(instr)
+
+    def visit_continue(self, instr: Continue) -> Optional[DataflowObject]:
+        return self._mapper(instr)
+
+    def visit_phi(self, instr: Phi) -> Optional[DataflowObject]:
+        # we ignore the return value here, because replacing instr.value itself would require updating
+        # instr.origin_block with information we don't have
+        instr.value.accept(self)
+
+        # update instr.origin_block with potential changes from instr.value.accept(self)
+        for node, expression in instr.origin_block.items():
+            if (replacement := self._mapper(expression)) is not None:
+                instr.origin_block[node] = _assert_type(replacement, Union[Variable, Constant])
+
+        if (destination_replacement := instr.destination.accept(self)) is not None:
+            instr._destination = _assert_type(destination_replacement, Variable)
+
+        return self._mapper(instr)
+
+    def visit_mem_phi(self, instr: MemPhi) -> Optional[DataflowObject]:
```

Comment:
I've decided to just plainly replace every encountered object in the dataflow tree, instead of doing it like the original `substitute` method with the reasoning explained in my other [comment](https://github.com/fkie-cad/dewolf/pull/318#discussion_r1317704755).